modules/stochastic_tools/include/utils/GaussianProcessHandler.h

```diff
@@ -121,9 +125,22 @@ class GaussianProcessHandler
   static PetscErrorCode
   formFunctionGradientWrapper(Tao tao, Vec theta, PetscReal * f, Vec Grad, void * ptr);
 
-  // Computes Gradient of the loss function
+  // Computes Gradient of the loss function for TAO usage
   void formFunctionGradient(Tao tao, Vec theta, PetscReal * f, Vec Grad);
 
+  // Tune hyperparameters using ADAM
+  void tuneHyperParamsADAM(RealEigenMatrix training_params,
+                           RealEigenMatrix training_data,
+                           const unsigned int & iter,
+                           const unsigned int & batch_size,
+                           const Real & learningRate);
+
+  // Computes the loss function for ADAM usage
+  Real getLossADAM();
```

Comment:
```suggestion
  Real getLossADAM(RealEigenMatrix & inputs, RealEigenMatrix & outputs);
```
I think the same could be done for the gradient estimation too. Considering that you already create a shuffled `input` and `output` matrix, you can use it directly instead of `_training_data` (which I'd like to keep intact).